File: Data Structures & Algorithms/capacity-to-ship-packages-within-d-days/submission-4.py
class Solution:
    def shipWithinDays(self, weights: List[int], days: int) -> int:
        left, right = max(weights), sum(weights)

        def can_ship(threshold):
            ship, curr_capacity = 1, threshold
            for weight in weights:
                if curr_capacity - weight < 0:
                    ship += 1
                    if ship > days:
                        return False
                    curr_capacity = threshold
                curr_capacity -= weight
            return True

        while left <= right:
            mid = left + (right - left) // 2

            if can_ship(mid):
                right = mid - 1
            else:
                left = mid + 1

        return left

File: Data Structures & Algorithms/sqrtx/submission-4.go
func mySqrt(x int) int {
    if x == 0 {
        return 0
    }

    left := 1
    right := x

    for left <= right {
        mid := left + (right-left)/2

        if mid * mid == x {
            return mid
        } else if mid * mid > x {
            right = mid - 1
        } else {
            left = mid + 1
        }
    }

    return right
}
